packages/context/src/binding.ts

```diff
@@ -94,6 +120,52 @@ export class Binding {
    * ```
    */
   getValue(ctx: Context): BoundValue | Promise<BoundValue> {
+    // First check cached value for non-transient
+    if (this._cache !== undefined) {
+      if (this._scope === BindingScope.SINGLETON) {
+        return this._cache;
+      } else if (this._scope === BindingScope.CONTEXT) {
+        if (ctx.contains(this._key)) {
+          return this._cache;
+        }
+      }
+    }
+    if (this._getValue) {
+      const result = this._getValue(ctx);
+      if (isPromise(result)) {
+        if (this._scope === BindingScope.SINGLETON) {
+          // Cache the value
+          result.then(val => {
```

Comment:
This does not look correct to me. Here is how I understand this code:

```ts
result.then(updateCache);
return result;
```

The `then` handler added here will basically run in parallel with whatever user-added `then` handler added to what `getValue` returns.

IMO, the correct solution is to return a promise that's resolved only after the cache has been updated.

```ts
result = result.then(val => {
  // update cache
});
```